web/src/stream_list.ts

```diff
@@ -822,13 +825,51 @@ export function set_event_handlers({
 }: {
     on_stream_click: (stream_id: number, trigger: string) => void;
 }): void {
-    $("#stream_filters").on("click", "li .subscription_block", (e) => {
+    $("#stream_filters").on(
+        "click",
+        "li .subscription_block .goto-interleaved-stream-view",
+        (e) => {
+            if (e.metaKey || e.ctrlKey) {
+                return;
+            }
+            const stream_id = stream_id_for_elt($(e.target).parents("li"));
+            on_stream_click(stream_id, "sidebar");
+
+            clear_and_hide_search();
+
+            e.preventDefault();
+            e.stopPropagation();
+        },
+    );
+
+    $("#stream_filters").on("click", "li .subscription_block .stream-name", (e) => {
         if (e.metaKey || e.ctrlKey) {
             return;
         }
+
         const stream_id = stream_id_for_elt($(e.target).parents("li"));
-        on_stream_click(stream_id, "sidebar");
+        const topics = stream_topic_history.get_recent_topic_names(stream_id);
+        let destination_url;
 
+        const top_unmuted_topic = topics.find(
+            (topic) => !user_topics.is_topic_muted(stream_id, topic),
+        );
+        const muted_topics = topics.find(
+            (topic) => !user_topics.is_topic_unmuted(stream_id, topic),
+        );
+
+        if (top_unmuted_topic) {
+            destination_url = hash_util.by_stream_topic_url(stream_id, top_unmuted_topic);
+            browser_history.go_to_location(destination_url);
+        } else if (muted_topics) {
+            destination_url = hash_util.by_stream_topic_url(stream_id, muted_topics);
+            browser_history.go_to_location(destination_url);
+        } else {
+            on_stream_click(stream_id, "sidebar");
+            return;
+        }
```

Comment:
So I think long-term, we're going to need to figure out better error handling for the case that we don't have anything in `stream_topic_history`; but going to the stream seems OK for testing.

Why do you `return` here, though?

Reply:
I think we don't need else block.

Reply:
No, as noted before, we need to go to the full-stream narrow if a stream contains no topics. I just reproduced it being broken with an empty stream.

Note that empty stream is not the only scenario where this may come up -- if we don't have any messages in the stream fetched in our local cache, this existing logic would also fail.

That second issue be patched by doing `stream_topic_history_util.get_server_history` with an appropriate callback, only for the case that we don't have data in `stream_topic_history.get_recent_topic_names`.

Reply:
I can test-deploy this with the following patch applied, also pushed it here.

``` diff
diff --git a/web/src/stream_list.ts b/web/src/stream_list.ts
index 06262de9dd..a14e07cb8f 100644
--- a/web/src/stream_list.ts
+++ b/web/src/stream_list.ts
@@ -876,6 +876,8 @@ export function set_event_handlers({
         } else if (muted_topics) {
             destination_url = hash_util.by_stream_topic_url(stream_id, muted_topics);
             browser_history.go_to_location(destination_url);
+        } else {
+            on_stream_click(stream_id, "sidebar");
         }
 
         popovers.hide_all();
```